// @paperclipai/server/dist/services/agent-secret-bindings.js
import { envBindingSchema } from "@paperclipai/shared";
function asRecord(value) {
    if (typeof value !== "object" || value === null || Array.isArray(value))
        return null;
    return value;
}
function collectSecretRefs(adapterConfig) {
    const config = asRecord(adapterConfig);
    if (!config)
        return [];
    const refs = [];
    const envValue = asRecord(config.env);
    for (const [key, rawBinding] of Object.entries(envValue ?? {})) {
        const parsed = envBindingSchema.safeParse(rawBinding);
        if (!parsed.success)
            continue;
        const binding = parsed.data;
        if (typeof binding !== "object" || binding === null || binding.type !== "secret_ref")
            continue;
        refs.push({
            secretId: binding.secretId,
            configPath: `env.${key}`,
            versionSelector: binding.version ?? "latest",
        });
    }
    for (const [key, rawBinding] of Object.entries(config)) {
        if (key === "env")
            continue;
        const parsed = envBindingSchema.safeParse(rawBinding);
        if (!parsed.success)
            continue;
        const binding = parsed.data;
        if (typeof binding !== "object" || binding === null || binding.type !== "secret_ref")
            continue;
        refs.push({
            secretId: binding.secretId,
            configPath: key,
            versionSelector: binding.version ?? "latest",
        });
    }
    return refs;
}
export async function syncAgentAdapterEnvBindings(input) {
    if (input.secretsSvc.syncSecretRefsForTarget) {
        await input.secretsSvc.syncSecretRefsForTarget(input.companyId, { targetType: "agent", targetId: input.agentId }, collectSecretRefs(input.adapterConfig), { replaceAll: true });
        return;
    }
    const envValue = asRecord(asRecord(input.adapterConfig)?.env);
    await input.secretsSvc.syncEnvBindingsForTarget?.(input.companyId, { targetType: "agent", targetId: input.agentId }, envValue);
}
