// @paperclipai/server/dist/services/environments.js
import { and, desc, eq, sql } from "drizzle-orm";
import { environmentLeases, environments } from "@paperclipai/db";
import { ENVIRONMENT_DRIVERS, ENVIRONMENT_LEASE_CLEANUP_STATUSES, ENVIRONMENT_LEASE_POLICIES, ENVIRONMENT_LEASE_STATUSES, ENVIRONMENT_STATUSES, } from "@paperclipai/shared";
import { conflict } from "../errors.js";
const DEFAULT_LOCAL_ENVIRONMENT_NAME = "Local";
const DEFAULT_LOCAL_ENVIRONMENT_DESCRIPTION = "Default execution environment for Paperclip runs on this machine.";
const DEFAULT_KUBERNETES_ENVIRONMENT_NAME = "Kubernetes Sandbox";
const DEFAULT_KUBERNETES_ENVIRONMENT_DESCRIPTION = "Managed Kubernetes sandbox environment for hosted tenant execution.";
/** Provider key (== plugin driverKey) of the first-party Kubernetes sandbox provider. */
const KUBERNETES_PROVIDER_KEY = "kubernetes";
/** Metadata marker for the company's managed-by-config Kubernetes sandbox environment. */
const KUBERNETES_MANAGED_MARKER = "managedKubernetesSandbox";
function cloneRecord(value, fallback = null) {
    if (!value || typeof value !== "object" || Array.isArray(value))
        return fallback;
    return { ...value };
}
function readEnum(value, allowed, fieldName) {
    if (value === null)
        return null;
    if (allowed.includes(value))
        return value;
    throw new Error(`Unexpected ${fieldName} value: ${value}`);
}
function hasConstraintName(error, constraintName) {
    if (typeof error !== "object" || error === null)
        return false;
    const candidate = error;
    return candidate.constraint === constraintName
        || candidate.constraint_name === constraintName
        || hasConstraintName(candidate.cause, constraintName);
}
function toEnvironment(row) {
    return {
        id: row.id,
        name: row.name,
        description: row.description ?? null,
        driver: readEnum(row.driver, ENVIRONMENT_DRIVERS, "environment driver") ?? "local",
        status: readEnum(row.status, ENVIRONMENT_STATUSES, "environment status") ?? "active",
        config: cloneRecord(row.config, {}) ?? {},
        envVars: cloneRecord(row.envVars, {}) ?? {},
        metadata: cloneRecord(row.metadata),
        createdAt: row.createdAt,
        updatedAt: row.updatedAt,
    };
}
function resolveListFilters(companyIdOrFilters, maybeFilters) {
    if (typeof companyIdOrFilters === "string") {
        return maybeFilters ?? {};
    }
    return companyIdOrFilters ?? {};
}
function resolveCreateInput(companyIdOrInput, maybeInput) {
    if (typeof companyIdOrInput === "string") {
        if (!maybeInput)
            throw new Error("Create environment input is required");
        return maybeInput;
    }
    return companyIdOrInput;
}
function resolveKubernetesConfig(companyIdOrConfig, maybeConfig) {
    if (typeof companyIdOrConfig === "string") {
        if (!maybeConfig)
            throw new Error("Kubernetes environment config is required");
        return maybeConfig;
    }
    return companyIdOrConfig;
}
function toEnvironmentLease(row) {
    return {
        id: row.id,
        companyId: row.companyId,
        environmentId: row.environmentId,
        executionWorkspaceId: row.executionWorkspaceId ?? null,
        issueId: row.issueId ?? null,
        heartbeatRunId: row.heartbeatRunId ?? null,
        status: readEnum(row.status, ENVIRONMENT_LEASE_STATUSES, "environment lease status") ?? "active",
        leasePolicy: readEnum(row.leasePolicy, ENVIRONMENT_LEASE_POLICIES, "environment lease policy") ?? "ephemeral",
        provider: row.provider ?? null,
        providerLeaseId: row.providerLeaseId ?? null,
        acquiredAt: row.acquiredAt,
        lastUsedAt: row.lastUsedAt,
        expiresAt: row.expiresAt ?? null,
        releasedAt: row.releasedAt ?? null,
        failureReason: row.failureReason ?? null,
        cleanupStatus: readEnum(row.cleanupStatus, ENVIRONMENT_LEASE_CLEANUP_STATUSES, "environment lease cleanup status"),
        metadata: cloneRecord(row.metadata),
        createdAt: row.createdAt,
        updatedAt: row.updatedAt,
    };
}
export function environmentService(db) {
    return {
        list: async (companyIdOrFilters, maybeFilters) => {
            const filters = resolveListFilters(companyIdOrFilters, maybeFilters);
            const conditions = [];
            if (filters.status)
                conditions.push(eq(environments.status, filters.status));
            if (filters.driver)
                conditions.push(eq(environments.driver, filters.driver));
            const rows = await db
                .select()
                .from(environments)
                .where(conditions.length > 0 ? and(...conditions) : undefined)
                .orderBy(desc(environments.updatedAt), desc(environments.createdAt));
            return rows.map(toEnvironment);
        },
        getById: async (id) => {
            const row = await db.select().from(environments).where(eq(environments.id, id)).then((rows) => rows[0] ?? null);
            return row ? toEnvironment(row) : null;
        },
        getLeaseById: async (id) => {
            const row = await db
                .select()
                .from(environmentLeases)
                .where(eq(environmentLeases.id, id))
                .then((rows) => rows[0] ?? null);
            return row ? toEnvironmentLease(row) : null;
        },
        ensureLocalEnvironment: async (_companyId) => {
            const now = new Date();
            const row = await db
                .insert(environments)
                .values({
                name: DEFAULT_LOCAL_ENVIRONMENT_NAME,
                description: DEFAULT_LOCAL_ENVIRONMENT_DESCRIPTION,
                driver: "local",
                status: "active",
                config: {},
                envVars: {},
                metadata: {
                    managedByPaperclip: true,
                    defaultForInstance: true,
                },
                createdAt: now,
                updatedAt: now,
            })
                .onConflictDoNothing({
                target: [environments.driver],
                where: sql `${environments.driver} = 'local'`,
            })
                .returning()
                .then((rows) => rows[0] ?? null);
            if (row)
                return toEnvironment(row);
            const existing = await db
                .select()
                .from(environments)
                .where(eq(environments.driver, "local"))
                .then((rows) => rows[0] ?? null);
            if (!existing) {
                throw new Error("Failed to ensure local environment");
            }
            return toEnvironment(existing);
        },
        /**
         * Idempotently ensure a managed Kubernetes sandbox environment exists for a
         * instance, configured from instance/operator-supplied config. Mirrors
         * `ensureLocalEnvironment`, but there is no DB unique index for sandbox
         * drivers, so idempotency is by metadata marker + driver lookup.
         *
         * The environment is `driver: "sandbox"` with `config.provider:
         * "kubernetes"` so it resolves to the first-party Kubernetes sandbox
         * provider. On subsequent calls the config is refreshed (so operators can
         * update egress/runtimeClass via gitops without recreating the row).
         */
        ensureKubernetesEnvironment: async (companyIdOrConfig, maybeConfig) => {
            const config = resolveKubernetesConfig(companyIdOrConfig, maybeConfig);
            const desiredConfig = {
                ...config,
                provider: KUBERNETES_PROVIDER_KEY,
            };
            const desiredMetadata = {
                managedByPaperclip: true,
                [KUBERNETES_MANAGED_MARKER]: true,
            };
            const existing = await db
                .select()
                .from(environments)
                .where(eq(environments.driver, "sandbox"))
                .then((rows) => rows.find((row) => row.metadata?.[KUBERNETES_MANAGED_MARKER] === true) ?? null);
            const now = new Date();
            if (existing) {
                const updated = await db
                    .update(environments)
                    .set({
                    config: desiredConfig,
                    metadata: { ...(existing.metadata ?? {}), ...desiredMetadata },
                    status: "active",
                    updatedAt: now,
                })
                    .where(eq(environments.id, existing.id))
                    .returning()
                    .then((rows) => rows[0] ?? existing);
                return toEnvironment(updated);
            }
            // The partial unique index `environments_managed_sandbox_idx` enforces
            // "at most one Paperclip-managed sandbox row per instance" at the DB
            // level. Use ON CONFLICT DO NOTHING keyed on that index so concurrent
            // callers can race the INSERT; losers re-read the surviving row.
            const inserted = await db
                .insert(environments)
                .values({
                name: DEFAULT_KUBERNETES_ENVIRONMENT_NAME,
                description: DEFAULT_KUBERNETES_ENVIRONMENT_DESCRIPTION,
                driver: "sandbox",
                status: "active",
                config: desiredConfig,
                envVars: {},
                metadata: desiredMetadata,
                createdAt: now,
                updatedAt: now,
            })
                .onConflictDoNothing({
                target: [environments.driver],
                where: sql `${environments.driver} = 'sandbox' AND (${environments.metadata} ->> 'managedByPaperclip')::boolean = true`,
            })
                .returning()
                .then((rows) => rows[0] ?? null)
                .catch((error) => {
                if (hasConstraintName(error, "environments_name_idx")
                    || hasConstraintName(error, "environments_managed_sandbox_idx")) {
                    return null;
                }
                throw error;
            });
            if (inserted)
                return toEnvironment(inserted);
            const winner = await db
                .select()
                .from(environments)
                .where(eq(environments.driver, "sandbox"))
                .then((rows) => rows.find((candidate) => candidate.metadata?.[KUBERNETES_MANAGED_MARKER] === true) ?? null);
            if (!winner) {
                throw new Error("Failed to ensure kubernetes environment");
            }
            return toEnvironment(winner);
        },
        /**
         * Find the active managed Kubernetes sandbox environment, if one
         * exists. Read-only counterpart to `ensureKubernetesEnvironment` used by the
         * per-run execution guard (which must not silently create config-less envs).
         */
        findKubernetesEnvironment: async (_companyId) => {
            const rows = await db
                .select()
                .from(environments)
                .where(and(eq(environments.driver, "sandbox"), eq(environments.status, "active")))
                .orderBy(desc(environments.updatedAt));
            const match = rows.find((row) => row.metadata?.[KUBERNETES_MANAGED_MARKER] === true);
            return match ? toEnvironment(match) : null;
        },
        create: async (companyIdOrInput, maybeInput) => {
            const input = resolveCreateInput(companyIdOrInput, maybeInput);
            const now = new Date();
            const row = await db
                .insert(environments)
                .values({
                name: input.name,
                description: input.description ?? null,
                driver: input.driver,
                status: input.status ?? "active",
                config: input.config ?? {},
                envVars: input.envVars ?? {},
                metadata: input.metadata ?? null,
                createdAt: now,
                updatedAt: now,
            })
                .returning()
                .then((rows) => rows[0] ?? null)
                .catch((error) => {
                if (hasConstraintName(error, "environments_name_idx")) {
                    throw conflict(`An environment named "${input.name}" already exists for this instance.`);
                }
                if (hasConstraintName(error, "environments_local_driver_idx")) {
                    throw conflict("A local environment already exists for this instance.");
                }
                throw error;
            });
            if (!row) {
                throw new Error("Failed to create environment");
            }
            return toEnvironment(row);
        },
        update: async (id, patch) => {
            const values = {
                updatedAt: new Date(),
            };
            if (patch.name !== undefined)
                values.name = patch.name;
            if (patch.description !== undefined)
                values.description = patch.description ?? null;
            if (patch.driver !== undefined)
                values.driver = patch.driver;
            if (patch.status !== undefined)
                values.status = patch.status;
            if (patch.config !== undefined)
                values.config = patch.config;
            if ("envVars" in patch && patch.envVars !== undefined) {
                values.envVars = (patch.envVars ?? {});
            }
            if (patch.metadata !== undefined)
                values.metadata = patch.metadata ?? null;
            const row = await db
                .update(environments)
                .set(values)
                .where(eq(environments.id, id))
                .returning()
                .then((rows) => rows[0] ?? null)
                .catch((error) => {
                if (hasConstraintName(error, "environments_name_idx")) {
                    throw conflict(`An environment named "${patch.name}" already exists for this instance.`);
                }
                if (hasConstraintName(error, "environments_local_driver_idx")) {
                    throw conflict("A local environment already exists for this instance.");
                }
                throw error;
            });
            return row ? toEnvironment(row) : null;
        },
        remove: async (id) => {
            const row = await db
                .delete(environments)
                .where(eq(environments.id, id))
                .returning()
                .then((rows) => rows[0] ?? null);
            return row ? toEnvironment(row) : null;
        },
        listLeases: async (environmentId, filters = {}) => {
            const conditions = [eq(environmentLeases.environmentId, environmentId)];
            if (filters.status)
                conditions.push(eq(environmentLeases.status, filters.status));
            const rows = await db
                .select()
                .from(environmentLeases)
                .where(and(...conditions))
                .orderBy(desc(environmentLeases.lastUsedAt), desc(environmentLeases.createdAt));
            return rows.map(toEnvironmentLease);
        },
        acquireLease: async (input) => {
            const now = new Date();
            const row = await db
                .insert(environmentLeases)
                .values({
                companyId: input.companyId,
                environmentId: input.environmentId,
                executionWorkspaceId: input.executionWorkspaceId ?? null,
                issueId: input.issueId ?? null,
                heartbeatRunId: input.heartbeatRunId ?? null,
                status: "active",
                leasePolicy: input.leasePolicy ?? "ephemeral",
                provider: input.provider ?? null,
                providerLeaseId: input.providerLeaseId ?? null,
                acquiredAt: now,
                lastUsedAt: now,
                expiresAt: input.expiresAt ?? null,
                releasedAt: null,
                failureReason: null,
                cleanupStatus: null,
                metadata: input.metadata ?? null,
                createdAt: now,
                updatedAt: now,
            })
                .returning()
                .then((rows) => rows[0] ?? null);
            if (!row) {
                throw new Error("Failed to acquire environment lease");
            }
            return toEnvironmentLease(row);
        },
        releaseLease: async (id, status = "released", options) => {
            const now = new Date();
            const row = await db
                .update(environmentLeases)
                .set({
                status,
                releasedAt: status === "retained" ? null : now,
                lastUsedAt: now,
                updatedAt: now,
                ...(options?.failureReason !== undefined ? { failureReason: options.failureReason } : {}),
                ...(options?.cleanupStatus !== undefined ? { cleanupStatus: options.cleanupStatus } : {}),
            })
                .where(eq(environmentLeases.id, id))
                .returning()
                .then((rows) => rows[0] ?? null);
            return row ? toEnvironmentLease(row) : null;
        },
        updateLeaseMetadata: async (id, metadata) => {
            const row = await db
                .update(environmentLeases)
                .set({
                metadata,
                lastUsedAt: new Date(),
                updatedAt: new Date(),
            })
                .where(eq(environmentLeases.id, id))
                .returning()
                .then((rows) => rows[0] ?? null);
            return row ? toEnvironmentLease(row) : null;
        },
        releaseLeasesForRun: async (heartbeatRunId, status = "released") => {
            const now = new Date();
            const rows = await db
                .update(environmentLeases)
                .set({
                status,
                releasedAt: now,
                lastUsedAt: now,
                updatedAt: now,
            })
                .where(and(eq(environmentLeases.heartbeatRunId, heartbeatRunId), eq(environmentLeases.status, "active")))
                .returning();
            return rows.map(toEnvironmentLease);
        },
    };
}
