// @paperclipai/server/dist/services/execution-allowlist.js
/**
 * Pure execution-allowlist guard.
 *
 * Decides whether a candidate execution environment is permitted to run an
 * agent, given the instance-level execution policy. This is security-critical:
 * on a shared cloud instance we FORCE all untrusted tenant agents onto the
 * Kubernetes sandbox-provider and REFUSE local/in-process execution so that a
 * tenant agent can never run inside the server process or on an unsandboxed
 * local/ssh adapter.
 *
 * The merged tree's environment model represents the Kubernetes sandbox as a
 * core `driver: "sandbox"` environment whose `config.provider` is the plugin's
 * `driverKey` ("kubernetes", `kind: "sandbox_provider"`). The local default is
 * `driver: "local"`. This module knows nothing about the DB or heartbeat — it
 * just maps (driver, provider, policy) -> allow/deny so it is trivially
 * unit-testable.
 */
/** Provider key (== plugin driverKey) of the first-party Kubernetes sandbox provider. */
export const KUBERNETES_PROVIDER_KEY = "kubernetes";
/** True when the policy forces all execution onto the Kubernetes sandbox. */
export function isExecutionForcedToKubernetes(policy) {
    return policy?.executionMode === "kubernetes";
}
/**
 * True iff the candidate environment is the Kubernetes sandbox provider, i.e. a
 * core `sandbox` driver whose provider key is "kubernetes".
 */
export function isKubernetesSandboxEnvironment(candidate) {
    return candidate.driver === "sandbox" && candidate.provider === KUBERNETES_PROVIDER_KEY;
}
/**
 * Decide whether the candidate environment may run under the given policy.
 *
 * When `executionMode === "kubernetes"`, ONLY a `sandbox_provider` driver with
 * provider/driverKey "kubernetes" is allowed; a `local` driver (or any non-k8s
 * sandbox provider, or ssh, or plugin) is DENIED. Otherwise everything is
 * allowed.
 */
export function evaluateExecutionAllowlist(policy, candidate) {
    if (!isExecutionForcedToKubernetes(policy)) {
        return { allowed: true };
    }
    if (isKubernetesSandboxEnvironment(candidate)) {
        return { allowed: true };
    }
    const provider = candidate.provider ?? null;
    const target = candidate.driver === "sandbox"
        ? `sandbox provider "${provider ?? "(none)"}"`
        : `"${candidate.driver}" driver`;
    return {
        allowed: false,
        reason: `Instance execution policy requires the Kubernetes sandbox provider ` +
            `(executionMode=kubernetes), but the resolved environment uses the ${target}. ` +
            `Untrusted execution on a non-Kubernetes environment is refused.`,
        deniedDriver: candidate.driver,
        deniedProvider: provider,
    };
}
