// @paperclipai/adapter-utils/dist/remote-managed-runtime.js
import path from "node:path";
import { GIT_ARCHIVE_EXCLUDES } from "./git-workspace-sync.js";
import { prepareWorkspaceForSshExecution, restoreWorkspaceFromSshExecution, syncDirectoryToSsh, } from "./ssh.js";
import { captureDirectorySnapshot } from "./workspace-restore-merge.js";
function asObject(value) {
    return value && typeof value === "object" && !Array.isArray(value)
        ? value
        : {};
}
function asString(value) {
    return typeof value === "string" ? value : "";
}
function asNumber(value) {
    return typeof value === "number" ? value : Number(value);
}
export function buildRemoteExecutionSessionIdentity(spec) {
    if (!spec)
        return null;
    return {
        transport: "ssh",
        host: spec.host,
        port: spec.port,
        username: spec.username,
        remoteCwd: spec.remoteCwd,
    };
}
export function remoteExecutionSessionMatches(saved, current) {
    const currentIdentity = buildRemoteExecutionSessionIdentity(current);
    if (!currentIdentity)
        return false;
    const parsedSaved = asObject(saved);
    return (asString(parsedSaved.transport) === currentIdentity.transport &&
        asString(parsedSaved.host) === currentIdentity.host &&
        asNumber(parsedSaved.port) === currentIdentity.port &&
        asString(parsedSaved.username) === currentIdentity.username &&
        asString(parsedSaved.remoteCwd) === currentIdentity.remoteCwd);
}
export async function prepareRemoteManagedRuntime(input) {
    const baseWorkspaceRemoteDir = input.workspaceRemoteDir ?? input.spec.remoteCwd;
    const workspaceRemoteDir = path.posix.join(baseWorkspaceRemoteDir, ".paperclip-runtime", "runs", input.runId, "workspace");
    const runtimeRootDir = path.posix.join(workspaceRemoteDir, ".paperclip-runtime", input.adapterKey);
    const preparedWorkspace = await prepareWorkspaceForSshExecution({
        spec: input.spec,
        localDir: input.workspaceLocalDir,
        remoteDir: workspaceRemoteDir,
        onProgress: input.onProgress,
    });
    const restoreExclude = preparedWorkspace.gitBacked ? [...GIT_ARCHIVE_EXCLUDES, ".paperclip-runtime"] : [".paperclip-runtime"];
    const baselineSnapshot = await captureDirectorySnapshot(input.workspaceLocalDir, {
        exclude: restoreExclude,
    });
    const assetDirs = {};
    try {
        for (const asset of input.assets ?? []) {
            const remoteDir = path.posix.join(runtimeRootDir, asset.key);
            assetDirs[asset.key] = remoteDir;
            await syncDirectoryToSsh({
                spec: input.spec,
                localDir: asset.localDir,
                remoteDir,
                followSymlinks: asset.followSymlinks,
                exclude: asset.exclude,
                onProgress: input.onProgress,
                progressLabel: asset.key,
            });
        }
    }
    catch (error) {
        await restoreWorkspaceFromSshExecution({
            spec: input.spec,
            localDir: input.workspaceLocalDir,
            remoteDir: workspaceRemoteDir,
            baselineSnapshot,
            restoreGitHistory: preparedWorkspace.gitBacked,
            onProgress: input.onProgress,
        });
        throw error;
    }
    return {
        spec: input.spec,
        workspaceLocalDir: input.workspaceLocalDir,
        workspaceRemoteDir,
        runtimeRootDir,
        assetDirs,
        restoreWorkspace: async (onProgress) => {
            await restoreWorkspaceFromSshExecution({
                spec: input.spec,
                localDir: input.workspaceLocalDir,
                remoteDir: workspaceRemoteDir,
                baselineSnapshot,
                restoreGitHistory: preparedWorkspace.gitBacked,
                onProgress,
            });
        },
    };
}
