// @paperclipai/adapter-utils/dist/runtime-progress.js
// Shared, throttled progress reporting for execution-target sync/restore.
//
// Transports (sandbox / SSH) own the byte counting and call `report()` as bytes
// move; orchestrators own the per-phase label and direction. The reporter
// throttles emits so a long transfer doesn't flood the log: a line is emitted
// only when the percentage crosses a step boundary (default every 10%) or once
// at least `minIntervalMs` has elapsed since the last emit. The terminal
// completion line is always emitted via `complete()` (or when `report()` reaches
// the known total).
const BYTES_PER_MB = 1024 * 1024;
function formatMb(bytes) {
    return (Math.max(0, bytes) / BYTES_PER_MB).toFixed(1);
}
function clampPercent(value) {
    if (!Number.isFinite(value))
        return 0;
    return Math.min(100, Math.max(0, Math.round(value)));
}
export function createRuntimeProgressReporter(options) {
    const stepPercent = options.stepPercent && options.stepPercent > 0 ? options.stepPercent : 10;
    const minIntervalMs = options.minIntervalMs && options.minIntervalMs > 0 ? options.minIntervalMs : 2000;
    const now = options.now ?? Date.now;
    const prefix = `[paperclip] ${options.phase}${options.label ? ` ${options.label}` : ""} ${options.direction} ${options.target}`;
    let lastEmitAt = null;
    let lastStep = -1;
    let lastDoneBytes = 0;
    let lastTotalBytes = null;
    let completed = false;
    function buildLine(doneBytes, totalBytes) {
        if (totalBytes != null && totalBytes > 0) {
            const pct = clampPercent((doneBytes / totalBytes) * 100);
            return `${prefix}: ${pct}% (${formatMb(doneBytes)}/${formatMb(totalBytes)} MB)\n`;
        }
        return `${prefix}: ${formatMb(doneBytes)} MB\n`;
    }
    function buildFailLine(doneBytes, totalBytes) {
        if (totalBytes != null && totalBytes > 0) {
            const pct = clampPercent((doneBytes / totalBytes) * 100);
            return `${prefix}: failed at ${pct}% (${formatMb(doneBytes)}/${formatMb(totalBytes)} MB)\n`;
        }
        return `${prefix}: failed after ${formatMb(doneBytes)} MB\n`;
    }
    async function emit(doneBytes, totalBytes) {
        lastEmitAt = now();
        if (totalBytes != null && totalBytes > 0) {
            lastStep = Math.floor(((doneBytes / totalBytes) * 100) / stepPercent);
        }
        await options.sink(buildLine(doneBytes, totalBytes));
    }
    return {
        async report(doneBytes, totalBytes) {
            lastDoneBytes = doneBytes;
            lastTotalBytes = totalBytes;
            if (completed)
                return;
            const elapsedOk = lastEmitAt == null || now() - lastEmitAt >= minIntervalMs;
            if (totalBytes != null && totalBytes > 0) {
                const terminal = doneBytes >= totalBytes;
                const step = Math.floor(((doneBytes / totalBytes) * 100) / stepPercent);
                const stepOk = step > lastStep;
                if (terminal || stepOk || elapsedOk) {
                    await emit(doneBytes, totalBytes);
                }
                if (terminal)
                    completed = true;
                return;
            }
            // Unknown total: no step boundaries, throttle purely on elapsed time.
            if (elapsedOk) {
                await emit(doneBytes, totalBytes);
            }
        },
        async complete(doneBytes, totalBytes) {
            if (completed)
                return;
            completed = true;
            const total = totalBytes !== undefined ? totalBytes : lastTotalBytes;
            const done = doneBytes !== undefined
                ? doneBytes
                : total != null && total > 0
                    ? total
                    : lastDoneBytes;
            await options.sink(buildLine(done, total));
        },
        async fail(doneBytes, totalBytes) {
            if (completed)
                return;
            completed = true;
            const total = totalBytes !== undefined ? totalBytes : lastTotalBytes;
            const done = doneBytes !== undefined ? doneBytes : lastDoneBytes;
            await options.sink(buildFailLine(done, total));
        },
    };
}
