// @agentclientprotocol/sdk/dist/connection.js
import { isResponseMessage } from "./jsonrpc.js";
import { messageIdKey, sessionIdFromMessageParams, sessionIdFromResponseResult, } from "./protocol.js";
export class OutboundStream {
    capacity;
    subscribers = new Set();
    replayBuffer = [];
    hasSubscriber = false;
    isClosed = false;
    constructor(capacity = 1024) {
        this.capacity = capacity;
    }
    push(message) {
        if (this.isClosed) {
            return;
        }
        if (!this.hasSubscriber) {
            this.replayBuffer.push(message);
            if (this.replayBuffer.length > this.capacity) {
                this.replayBuffer.shift();
            }
            return;
        }
        for (const subscriber of this.subscribers) {
            subscriber.push(message);
        }
    }
    subscribe() {
        const replay = this.hasSubscriber ? [] : [...this.replayBuffer];
        this.replayBuffer = [];
        this.hasSubscriber = true;
        const subscriber = new OutboundSubscriber(this.capacity, (item) => {
            this.subscribers.delete(item);
        });
        this.subscribers.add(subscriber);
        if (this.isClosed) {
            subscriber.close();
        }
        return {
            replay,
            stream: subscriber.stream,
        };
    }
    close() {
        if (this.isClosed) {
            return;
        }
        this.isClosed = true;
        this.replayBuffer = [];
        for (const subscriber of this.subscribers) {
            subscriber.close();
        }
        this.subscribers.clear();
    }
}
export class ConnectionState {
    connectionId;
    inboundTx;
    outboundRx;
    connectionStream = new OutboundStream();
    allOutbound = new OutboundStream();
    sessionStreams = new Map();
    pendingRoutes = new Map();
    clientResponseRoutes = new Map();
    closed;
    agentConnection;
    hasStartedRouter = false;
    inboundWriteChain = Promise.resolve();
    initialReader;
    outboundReader;
    shutdownPromise;
    resolveClosed = () => { };
    constructor(agent) {
        this.connectionId = globalThis.crypto.randomUUID();
        this.closed = new Promise((resolve) => {
            this.resolveClosed = resolve;
        });
        const inbound = new TransformStream();
        const outbound = new TransformStream();
        this.inboundTx = inbound.writable;
        this.outboundRx = outbound.readable;
        const stream = {
            readable: inbound.readable,
            writable: outbound.writable,
        };
        this.agentConnection = agent.connect(stream, {
            deferConnectHandlers: true,
        });
        this.observeAgentConnection();
    }
    async recvInitial(initializeId) {
        const reader = this.outboundRx.getReader();
        this.initialReader = reader;
        try {
            const result = await reader.read();
            if (result.done ||
                !result.value ||
                !isMatchingResponse(result.value, initializeId)) {
                if (!this.shutdownPromise) {
                    await this.shutdown();
                }
                throw new Error("Expected initialize response from agent");
            }
            return result.value;
        }
        finally {
            if (this.initialReader === reader) {
                this.initialReader = undefined;
            }
            reader.releaseLock();
        }
    }
    async writeInbound(message) {
        const write = this.inboundWriteChain.then(() => this.writeInboundMessage(message));
        this.inboundWriteChain = write.catch(() => undefined);
        await write;
    }
    startRouter() {
        if (this.hasStartedRouter) {
            return;
        }
        this.hasStartedRouter = true;
        void this.runRouter();
    }
    startConnectHandlers() {
        if (typeof this.agentConnection === "object" &&
            this.agentConnection !== null &&
            "startConnectHandlers" in this.agentConnection &&
            typeof this.agentConnection.startConnectHandlers === "function") {
            this.agentConnection.startConnectHandlers();
        }
    }
    ensureSession(sessionId) {
        const existing = this.sessionStreams.get(sessionId);
        if (existing) {
            return existing;
        }
        const stream = new OutboundStream();
        this.sessionStreams.set(sessionId, stream);
        return stream;
    }
    async shutdown() {
        if (!this.shutdownPromise) {
            this.shutdownPromise = this.runShutdown();
        }
        return this.shutdownPromise;
    }
    async runShutdown() {
        try {
            this.connectionStream.close();
            this.allOutbound.close();
            for (const stream of this.sessionStreams.values()) {
                stream.close();
            }
            this.sessionStreams.clear();
            this.pendingRoutes.clear();
            this.clientResponseRoutes.clear();
            await Promise.allSettled([
                this.inboundTx.close(),
                this.cancelOutboundReader(),
            ]);
        }
        finally {
            this.resolveClosed();
        }
    }
    observeAgentConnection() {
        if (typeof this.agentConnection !== "object" ||
            this.agentConnection === null ||
            !("closed" in this.agentConnection) ||
            !this.agentConnection.closed) {
            return;
        }
        void Promise.resolve(this.agentConnection.closed).finally(() => {
            void this.shutdown();
        });
    }
    cancelOutboundReader() {
        const reader = this.initialReader ?? this.outboundReader;
        if (reader) {
            return reader.cancel();
        }
        return this.outboundRx.cancel();
    }
    async writeInboundMessage(message) {
        const writer = this.inboundTx.getWriter();
        try {
            await writer.write(message);
        }
        finally {
            writer.releaseLock();
        }
    }
    async runRouter() {
        const reader = this.outboundRx.getReader();
        this.outboundReader = reader;
        try {
            while (true) {
                const result = await reader.read();
                if (result.done) {
                    return;
                }
                this.routeOutbound(result.value);
            }
        }
        catch (error) {
            console.error("ACP connection router stopped unexpectedly:", error);
        }
        finally {
            if (this.outboundReader === reader) {
                this.outboundReader = undefined;
            }
            reader.releaseLock();
            this.connectionStream.close();
            this.allOutbound.close();
            for (const stream of this.sessionStreams.values()) {
                stream.close();
            }
        }
    }
    routeOutbound(message) {
        this.allOutbound.push(message);
        if (isResponseMessage(message)) {
            this.routeOutboundResponse(message);
            return;
        }
        this.routeOutboundRequestOrNotification(message);
    }
    routeOutboundResponse(message) {
        const key = messageIdKey(message.id);
        const route = key ? this.pendingRoutes.get(key) : undefined;
        const sessionId = sessionIdFromResponseResult(message);
        if (sessionId) {
            this.ensureSession(sessionId);
        }
        if (key) {
            this.pendingRoutes.delete(key);
        }
        this.pushToRoute(route ?? "connection", message);
    }
    routeOutboundRequestOrNotification(message) {
        const sessionId = sessionIdFromMessageParams(message);
        if (sessionId) {
            this.trackClientResponseRoute(message, { session: sessionId });
            this.ensureSession(sessionId).push(message);
            return;
        }
        this.trackClientResponseRoute(message, "connection");
        this.connectionStream.push(message);
    }
    trackClientResponseRoute(message, route) {
        if (!("id" in message) || !("method" in message)) {
            return;
        }
        const key = messageIdKey(message.id);
        if (key) {
            this.clientResponseRoutes.set(key, route);
        }
    }
    pushToRoute(route, message) {
        if (route === "connection") {
            this.connectionStream.push(message);
            return;
        }
        this.ensureSession(route.session).push(message);
    }
}
export class ConnectionRegistry {
    connections = new Map();
    pendingConnections = new Map();
    createConnection(agent) {
        const connection = new ConnectionState(agent);
        this.connections.set(connection.connectionId, connection);
        this.trackConnectionClose(connection);
        return connection;
    }
    createPendingConnection(agent) {
        const connection = new ConnectionState(agent);
        this.pendingConnections.set(connection.connectionId, connection);
        this.trackConnectionClose(connection);
        return connection;
    }
    register(connection) {
        this.pendingConnections.delete(connection.connectionId);
        this.connections.set(connection.connectionId, connection);
    }
    get(connectionId) {
        return this.connections.get(connectionId);
    }
    remove(connectionId) {
        const connection = this.get(connectionId);
        if (!connection) {
            return undefined;
        }
        this.connections.delete(connectionId);
        void connection.shutdown();
        return connection;
    }
    discard(connectionId) {
        const connection = this.connections.get(connectionId) ??
            this.pendingConnections.get(connectionId);
        if (!connection) {
            return undefined;
        }
        this.connections.delete(connectionId);
        this.pendingConnections.delete(connectionId);
        void connection.shutdown();
        return connection;
    }
    async closeAll() {
        const connections = new Set([
            ...this.connections.values(),
            ...this.pendingConnections.values(),
        ]);
        this.connections.clear();
        this.pendingConnections.clear();
        await Promise.all(Array.from(connections, (connection) => connection.shutdown()));
    }
    trackConnectionClose(connection) {
        void connection.closed.then(() => {
            if (this.connections.get(connection.connectionId) === connection) {
                this.connections.delete(connection.connectionId);
            }
            if (this.pendingConnections.get(connection.connectionId) === connection) {
                this.pendingConnections.delete(connection.connectionId);
            }
        });
    }
}
class OutboundSubscriber {
    capacity;
    onCancel;
    stream;
    controller;
    queue = [];
    isClosed = false;
    hasWarnedAboutOverflow = false;
    constructor(capacity, onCancel) {
        this.capacity = capacity;
        this.onCancel = onCancel;
        this.stream = new ReadableStream({
            start: (controller) => {
                this.controller = controller;
                this.flush();
            },
            pull: () => {
                this.flush();
            },
            cancel: () => {
                this.cancel();
            },
        });
    }
    push(message) {
        if (this.isClosed) {
            return;
        }
        this.queue.push(message);
        if (this.queue.length > this.capacity) {
            this.queue.shift();
            if (!this.hasWarnedAboutOverflow) {
                console.warn("ACP outbound subscriber lagged; dropping oldest message");
                this.hasWarnedAboutOverflow = true;
            }
        }
        this.flush();
    }
    close() {
        if (this.isClosed) {
            return;
        }
        this.isClosed = true;
        this.queue = [];
        this.controller?.close();
    }
    cancel() {
        this.isClosed = true;
        this.queue = [];
        this.onCancel(this);
    }
    flush() {
        if (!this.controller) {
            return;
        }
        while (this.queue.length > 0 &&
            this.controller.desiredSize !== null &&
            this.controller.desiredSize > 0) {
            const message = this.queue.shift();
            if (!message) {
                return;
            }
            this.controller.enqueue(message);
        }
        if (this.queue.length === 0) {
            this.hasWarnedAboutOverflow = false;
        }
    }
}
function isMatchingResponse(msg, id) {
    return "id" in msg && !("method" in msg) && msg.id === id;
}
