// @paperclipai/server/dist/services/recovery/model-profile-hint.js
export const RECOVERY_MODEL_PROFILE_KEY = "cheap";
export const STATUS_ONLY_RECOVERY_GUARD_CONTEXT = {
    recoveryIntent: "status_only",
    allowDeliverableWork: false,
    allowDocumentUpdates: false,
    resumeRequiresNormalModel: true,
};
const RECOVERY_MODEL_PROFILE_HINT_KEYS = [
    "modelProfile",
    "paperclipModelProfile",
    "recoveryIntent",
    "allowDeliverableWork",
    "allowDocumentUpdates",
    "resumeRequiresNormalModel",
];
export function scrubRecoveryModelProfileHints(input) {
    const output = { ...input };
    for (const key of RECOVERY_MODEL_PROFILE_HINT_KEYS) {
        delete output[key];
    }
    return output;
}
export function withRecoveryModelProfileHint(input, workClass) {
    if (workClass === "normal_model") {
        return scrubRecoveryModelProfileHints(input);
    }
    return {
        ...scrubRecoveryModelProfileHints(input),
        ...STATUS_ONLY_RECOVERY_GUARD_CONTEXT,
        modelProfile: RECOVERY_MODEL_PROFILE_KEY,
    };
}
export function recoveryAssigneeAdapterOverrides(_workClass) {
    return { modelProfile: RECOVERY_MODEL_PROFILE_KEY };
}
